### PR TITLE
[WIP] Generalize xdg-shell popups and add to layer-shell

### PR DESCRIPTION
**Pull request by SirCmpwn**
* [x] Server implementation
* [x] Rendering
  * [ ] At the correct depth
  * [x] At the correct position
* [ ] Damage
* [ ] Grabs
* [x] Example client

### CODE REVIEW

**Comment by agx** [comment]
The popup parts that are outside the layer-shell in examples/layer-shell are not damaged properly but that's I guess that's expected since the todo list lacks a tick at "damage". Looks good otherwise!

File: types/wlr_layer_shell.c [hunk 1]
```diff
@@ -435,3 +448,74 @@ void wlr_layer_shell_destroy(struct wlr_layer_shell *layer_shell) {
...
+
+static void layer_surface_iterator(struct wlr_surface *surface,
+		int sx, int sy, void *data) {
+	struct layer_surface_iterator_data *iter_data = data;
+	iter_data->user_iterator(surface, iter_data->x + sx, iter_data->y + sy,
+		iter_data->user_data);
+}
+
+static void xdg_surface_for_each_surface(struct wlr_xdg_surface *surface,
+		int x, int y, wlr_surface_iterator_func_t iterator, void *user_data) {
```
Comment (emersion) [inline types/wlr_layer_shell.c]: You can use `wlr_xdg_surface_for_each_surface` instead of copy-pasting it here.

Comment (SirCmpwn) [reply types/wlr_layer_shell.c]: I had issues with that

Comment (emersion) [reply types/wlr_layer_shell.c]: It shouldn't be hard, you can use the `data` struct you defined earlier

Comment (SirCmpwn) [reply types/wlr_layer_shell.c]: That wasn't the issue. It was related to double-rendering popups

**Comment by SirCmpwn** [comment]
Closing in favor of https://github.com/swaywm/wlroots/pull/877

**Pull request closed without merging**